python/ray/train/callbacks/results_preprocessors/aggregate_fn.py

```diff
@@ -0,0 +1,104 @@
+import logging
+from typing import Dict, List, Union, Tuple, Optional
+
+import numpy as np
+
+VALID_AGGREGATE_TYPES: Tuple[type] = (
+    int,
+    float,
+    np.float32,
+    np.float64,
+    np.int32,
+    np.int64,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def _get_values_from_results(key, reported_metrics, results, get_weights=False):
+    """Return values in the results list from all workers."""
```

Comment:
What this function does is to retrieve values from `results_list`. But the behaviors of metric keys and weight keys are different:
1. For metric keys, if the metrics is not reported or its value type (from at least one worker) is invalid, this metric will be ignored, and a warning will be logged. (This may answer your [question](https://github.com/ray-project/ray/pull/22099#pullrequestreview-897929012).)
2. But for weight keys, if the weight value type is missing or invalid, then equal weight is used.

The idea behind this is based on the assumption that users may report incorrect or invalid key-value pairs in `train.report()`. But, as you said, the functionality here is complicated.

Reply:
Got it, thanks for the explanation! Can you please add this info to the docstring?

Also, since the behavior between metrics keys and weight keys are pretty different, I would recommend just using 2 different functions for them rather than using one function and having conditionals for the `get_weight` arg.